### PR TITLE
layers: Simplify Sampler state objects

### PR DESCRIPTION
**Pull request by jeremyg-lunarg**

### CODE REVIEW

File: layers/buffer_validation.cpp [hunk 1]
```diff
@@ -1706,10 +1706,9 @@ bool CoreChecks::ValidateCreateImageViewANDROID(const VkImageViewCreateInfo *cre
         uint64_t external_format = 0;
         const VkSamplerYcbcrConversionInfo *ycbcr_conv_info = LvlFindInChain<VkSamplerYcbcrConversionInfo>(create_info->pNext);
         if (ycbcr_conv_info != nullptr) {
-            VkSamplerYcbcrConversion conv_handle = ycbcr_conv_info->conversion;
-            if (ycbcr_conversion_ahb_fmt_map.find(conv_handle) != ycbcr_conversion_ahb_fmt_map.end()) {
-                conv_found = true;
```
Comment (filnet) [inline layers/buffer_validation.cpp:1711]: looks like this line should not be removed

Comment (jeremyg-lunarg) [reply layers/buffer_validation.cpp:1711]: Yes I think you're right. Thanks for catching that.

File: layers/state_tracker.cpp [hunk 2]
```diff
@@ -185,9 +165,6 @@ void ValidationStateTracker::PostCallRecordCreateImage(VkDevice device, const Vk
     if (VK_SUCCESS != result) return;
     auto is_node = std::make_shared<IMAGE_STATE>(device, *pImage, pCreateInfo);
     is_node->disjoint = ((pCreateInfo->flags & VK_IMAGE_CREATE_DISJOINT_BIT) != 0);
-    if (device_extensions.vk_android_external_memory_android_hardware_buffer) {
-        RecordCreateImageANDROID(pCreateInfo, is_node.get());
-    }
```
Comment (jeremyg-lunarg) [inline layers/state_tracker.cpp:190]: Still need to set format_features here.  Setting it in the IMAGE_STATE constructor is coming in a future patch.

File: layers/buffer_validation.cpp [hunk 3]
```diff
@@ -1706,10 +1706,10 @@ bool CoreChecks::ValidateCreateImageViewANDROID(const VkImageViewCreateInfo *cre
         uint64_t external_format = 0;
         const VkSamplerYcbcrConversionInfo *ycbcr_conv_info = LvlFindInChain<VkSamplerYcbcrConversionInfo>(create_info->pNext);
         if (ycbcr_conv_info != nullptr) {
-            VkSamplerYcbcrConversion conv_handle = ycbcr_conv_info->conversion;
-            if (ycbcr_conversion_ahb_fmt_map.find(conv_handle) != ycbcr_conversion_ahb_fmt_map.end()) {
+            auto ycbcr_state = GetSamplerYcbcrConversionState(ycbcr_conv_info->conversion);
```
Comment (ncesario-lunarg) [inline layers/buffer_validation.cpp:1709]: `const`?

File: layers/sampler_state.h [hunk 4]
```diff
@@ -50,33 +50,55 @@ struct less<SamplerUsedByImage> {
...
+    const VkSamplerCustomBorderColorCreateInfoEXT customCreateInfo;
 
     SAMPLER_STATE(const VkSampler *ps, const VkSamplerCreateInfo *pci)
-        : BASE_NODE(*ps, kVulkanObjectTypeSampler), createInfo(*pci) {
+        : BASE_NODE(*ps, kVulkanObjectTypeSampler),
+          createInfo(*pci),
+          samplerConversion(GetConversion(pci)),
+          customCreateInfo(GetCustomCreateInfo(pci)) {}
+
+    VkSampler sampler() const { return handle_.Cast<VkSampler>(); }
```
Comment (ncesario-lunarg) [inline layers/sampler_state.h:63]: Function name should probably be capitalized to stay within the style guidelines.

**Review by ncesario-lunarg — APPROVED** [review]
Couple minor style nits, but LGTM!

File: layers/sampler_state.h (hunk 4 above)
Comment (jeremyg-lunarg) [reply layers/sampler_state.h:63]: This method just got moved. I'd like to clean up all of the handle accessors in the state objects using some sort of C++ voodoo in a separate patch.

Comment (ncesario-lunarg) [reply layers/sampler_state.h:63]: Sorry, nvm. Missed that this just got moved. This probably fits under the "future refactor" department :)

**Pull request merged**